test/common/tls/integration/ssl_integration_test.cc

```diff
@@ -879,6 +934,90 @@ TEST_P(SslCertficateIntegrationTest, ServerRsaEcdsaClientEcdsaOnly) {
   checkStats();
 }
 
+// Server has RSA and ECDSA P-256 certificates, client only supports
+// P-384 curve. We fall back to RSA certificate.
+TEST_P(SslCertficateIntegrationTest, ServerRsaServerEcdsaP256EcdsaClientEcdsaP384Only) {
+  server_rsa_cert_ = true;
+  server_ecdsa_cert_ = true;
+  // We have to set the server curves to allow for ECDH negotiation where client
+  // doesn't support P-256
+  server_curves_.push_back("P-256");
```

Comment:
Can you elaborate on why this is needed? I don't understand. I would have guessed that setting the signature algorithms on the client would be enough, and leave the curves as the default.

Reply:
This appears to be required as the only time that the curves are set for the contexts is in this line, which uses the [ECDH settings to set the supported curves](https://github.com/envoyproxy/envoy/blob/93d05d5e69423ef051aa36da30b7568e22a0a0fc/source/common/tls/context_impl.cc#L148). As the defaults enabled are X25519 and P-256, we don't get a P-384+ context if we don't explicitly enable them on the server.

Reply:
Technically, the P-256 curve isn't required - the test would pass with P-384 only enabled, but we do need to explicitly enable it.

Reply:
But the curves are used for ECDH to generate ephemeral keys; I don't think they should matter for certificate selection and validation. And for that, I believe only the signature algorithms are used.

Reply:
I think I may have been thinking of something else here: our client for this test explicitly does *not* support P-256 for ECDH negotiation. We have to add the P-384 curve for the purposes of ECDH so that the client has a curve that it agrees on with the server to be able to do the ECDH negotiation. My earlier point about the defaults giving us only P-256 and X25519 curves available for that purpose stands.

I'll validate if the other tests pass without the specification of specific server curves.

Reply:
At least in tls 1.3, I don't think you should be setting supported curves on either client or server, you should be setting the signing algorithms on the client for what we're trying to test. I'm not sure if tls 1.2 co-mingles these or not.

Reply:
They do: the client _without_ P-256 support does, however, require the specification of a server-side curve that is does support.